zebra-state/src/service/read/tree.rs

```diff
@@ -164,96 +91,108 @@ where
 ///
-/// # Correctness
-///
-/// 1. After `chain` was cloned, the StateService can commit additional blocks to the finalized
-/// state `db`. Usually, the subtrees of these blocks are consistent. But if the `chain` is a
-/// different fork to `db`, then the trees can be inconsistent. In that case, we ignore all the
-/// trees in `chain` after the first inconsistent tree, because we know they will be inconsistent as
-/// well. (It is cryptographically impossible for tree roots to be equal once the leaves have
-/// diverged.)
-/// 2. APIs that return single subtrees can't be used here, because they can create
-/// an inconsistent list of subtrees after concurrent non-finalized and finalized updates.
+/// See [`subtrees`] for more details.
 pub fn orchard_subtrees<C>(
     chain: Option<C>,
     db: &ZebraDb,
-    start_index: NoteCommitmentSubtreeIndex,
-    limit: Option<NoteCommitmentSubtreeIndex>,
+    range: impl std::ops::RangeBounds<NoteCommitmentSubtreeIndex> + Clone,
 ) -> BTreeMap<NoteCommitmentSubtreeIndex, NoteCommitmentSubtreeData<orchard::tree::Node>>
 where
     C: AsRef<Chain>,
 {
-    let mut db_list = db.orchard_subtree_list_by_index_for_rpc(start_index, limit);
-
-    if let Some(limit) = limit {
-        let subtrees_num = u16::try_from(db_list.len())
-            .expect("There can't be more than `u16::MAX` Orchard subtrees.");
-
-        // Return the subtrees if the amount of them reached the given limit.
-        if subtrees_num == limit.0 {
-            return db_list;
-        }
-
-        // If not, make sure the amount is below the limit.
-        debug_assert!(subtrees_num < limit.0);
-    }
-
-    // If there's no chain, then we have the complete list.
-    let Some(chain) = chain else {
-        return db_list;
-    };
+    subtrees(
+        chain,
+        range,
+        |chain, range| chain.orchard_subtrees_in_range(range),
+        |range| db.orchard_subtree_list_by_index_range(range),
+    )
+}
 
-    // Unlike the other methods, this returns any trees in the range,
-    // even if there is no tree for start_index.
-    let fork_list = chain.as_ref().orchard_subtrees_in_range(start_index, limit);
+/// Returns a list of [`NoteCommitmentSubtree`]s in the provided range.
+///
+/// If there is no subtree at the first index in the range, the returned list is empty.
+/// Otherwise, subtrees are continuous up to the finalized tip.
+///
+/// Accepts a `chain` from the non-finalized state, a `range` of subtree indexes to retrieve,
+/// a `read_chain` function for retrieving the `range` of subtrees from `chain`, and
+/// a `read_disk` function for retrieving the `range` from [`ZebraDb`].
+///
+/// Returns without calling `read_disk` if `read_chain` returns a subtree at the first subtree index
+/// in the provided `range`. Otherwise, if `read_chain` and `read_disk` return different subtrees for
+/// the same subtree index, ignores all the trees in `chain` after the first inconsistent tree.
+///
```

Comment:
Should we move the correctness comment back up instead, or do would it help to add:

```suggestion
/// See correctness comment in function body for more details.
///
```

?

Reply:
I don't think we should duplicate this level of detail.

How about a summary of the goals in the function docs, and the details of how that is achieved inside the function in the existing detailed comment?

For example, the function doc could say:
> Returns a consistent set of subtrees for the supplied chain fork and database.
> Avoids reading the database if the trees are present in memory.

Reply:
Oh that's much better, applied in https://github.com/ZcashFoundation/zebra/pull/7739/commits/17c1a716505bf14904933cc09c26151d4a9e461e.